### ISSUE TITLE
Matrix multiplication in loop gives different output than batched

### ISSUE THREAD

**Issue by jbohnslav** [pytorch/pytorch#41248]
## 🐛 Bug

In my neural network architecture, I found that running inference on multiple images and then concatenating the results gives different results than batching them. The mean absolute error between the results is somewhere between 10^-6 and 10^-5. This seems large for floating point issues, but I could be wrong. Note: this has nothing to do with batchnorm, eval mode, etc, it's in `tensor.matmul`, as the below example shows. 

## To Reproduce

```python
import torch
import numpy as np
# reproducibility
torch.manual_seed(0)
np.random.seed(0)
print(torch.__version__)

# generate n rows of 512 feature vector and weights to go to 5 bins
num_iter = 10
input_np = np.random.standard_normal((num_iter, 512)).astype('float32')
weight_np = np.random.standard_normal((5, 512)).astype('float32')
# to torch
input = torch.from_numpy(input_np)
weight = torch.from_numpy(weight_np)
# matrix operation
output = input.matmul(weight.t()).cpu().numpy()
# loooop
output_loop = np.zeros(output.shape, dtype='float32')
for i in range(num_iter):
    input_l = torch.from_numpy(input_np[i, :])
    output_loop[i, :] = input_l.matmul(weight.t()).cpu().numpy()
# difference
print(f"Mean error between two methods: {np.mean(np.abs(output_loop - output))}")
```

## Expected behavior
I expect the mean absolute error between these two methods to be ~0. For the example above, it's ~4 * 10^-6. 

## Environment

```
PyTorch version: 1.7.0.dev20200709
Is debug build: No
CUDA used to build PyTorch: 10.1

OS: Ubuntu 18.04.4 LTS
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
CMake version: version 3.10.2

Python version: 3.7
Is CUDA available: Yes
CUDA runtime version: 10.1.243
GPU models and configuration: GPU 0: GeForce RTX 2080
Nvidia driver version: 418.87.00
cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.5

Versions of relevant libraries:
[pip3] numpy==1.19.0
[pip3] torch==1.7.0.dev20200709
[pip3] torchvision==0.8.0.dev20200710
[conda] blas                      1.0                         mkl  
[conda] cudatoolkit               10.1.243             h6bb024c_0  
[conda] mkl                       2020.1                      217  
[conda] mkl-service               2.3.0            py37he904b0f_0  
[conda] mkl_fft                   1.1.0            py37h23d657b_0  
[conda] mkl_random                1.1.1            py37h0573a6f_0  
[conda] numpy                     1.19.0                   pypi_0    pypi
[conda] numpy-base                1.18.5           py37hde5b4d6_0  
[conda] pytorch                   1.7.0.dev20200709 py3.7_cuda10.1.243_cudnn7.6.3_0    pytorch-nightly
[conda] torchvision               0.8.0.dev20200710      py37_cu101    pytorch-nightly
```

**Comment by SsnL** [comment]
for float32, error on the order of 10^-5 is expected.

**Comment by jbohnslav** [comment]
Okay, thank you!
